### PR TITLE
Product Collection: Implement the Hand-Picked collection

### LINKED ISSUE #50862: Product Collection - New "Hand-Picked" Collection
This task is about implementing Hand-Picked collection in Product Collection block.

Design: zx83mG5ZnDiacPdH6GlO2j-fi-2950_843243

image

**Scope**
- Implement Hand-Picked collection in Product Collection block.
- "Products to show" control in inspector control at the top.
  - This control looks similar to "Hand-picked products" filter control. Probably we can reuse the same control and move it to the top?
- "As entered" sorting option.
  - We decided to hide "sort by" control for Hand-Picked collection. Therefore, we will implement "As entered" sorting option as internal logic. This means that products will be displayed in the order they were added in the Hand-Picked control. For those, who want to have hand-picked products with "sort by" control, they can use custom Product Collection block with "Hand-picked products" filter. (For more info: (p1724312078913329-slack-C05UMTA7W0P))

### PR DESCRIPTION
**Pull request by WunderBart**
### Changes proposed in this Pull Request:

Closes #50862

This PR adds the Hand-Picked collection to the Product Collection block.

| Picker | Initial (empty) state | w/ picked products | w/ On Sale filter |
| ------------- | ------------- | ------------- | ------------- |
| Screenshot 2024-09-13 at 14 21 45 | image | Screenshot 2024-09-13 at 14 23 09 | Screenshot 2024-09-13 at 14 23 22 |

### How to test the changes in this Pull Request:

> [!IMPORTANT]
> See [Product Collection: Update heading for Hand-Picked collection](https://github.com/woocommerce/woocommerce/pull/51643#top) for updated testing instructions.

- Create a new post.
- Add the Product Collection block and select the Hand-Picked collection.
- Ensure the initial/empty state of the block displays `No products chosen yet. Use "Hand-Picked Products" in Inspector Controls to choose products to display`.
- Ensure the hand-picked control is visible at the top of the block settings panel (see image above).
- Ensure the hand-picked control is unavailable via the `Filters` setting dropdown, when no products are selected.
	- Save the draft and ensure the empty block is not rendered in the front end.
- Add some products in random order via the hand-picked control.
	- Ensure the "as entered" order is preserved in both editor and front end.
- Apply other filters (e.g., on sale) and ensure the results are as expected.

### Changelog entry

-   [x] Automatically create a changelog entry from the details below.

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>
<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [x] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [x] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 
Product Collection: Implement the Hand-Picked collection
</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce-blocks/assets/js/blocks/product-collection/collections/hand-picked.tsx
Comment (WunderBart) [inline plugins/woocommerce-blocks/assets/js/blocks/product-collection/collections/hand-picked.tsx]: The SVG here breaks the syntax highlight in GH, although it doesn't seem to break the code. Anyway, it doesn't feel right as is, but I'm not sure how to handle SVGs in this project since this seems to be the first one that is not available from the [@wordpress/icons package](https://wordpress.github.io/gutenberg/?path=/story/icons-icon--library):

Screenshot 2024-09-06 at 13 18 30


Any ideas, @imanish003 @kmanijak @xristos3490?

File: plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/hand-picked-products-control.tsx
Comment (WunderBart) [inline plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/hand-picked-products-control.tsx]: When the control is displayed as a part of filters, the label should be `Hand-picked products`, and when it's used as the Hand-picked collection-specific control, it should be `Select products`. What would be the best way to contextualize this label? Should we just pass an extra prop?

cc: @imanish003 @kmanijak @xristos3490

File: plugins/woocommerce-blocks/assets/js/blocks/product-collection/collections/hand-picked.tsx
Comment (imanish003) [reply plugins/woocommerce-blocks/assets/js/blocks/product-collection/collections/hand-picked.tsx]: Maybe we can use the same approach as in the [Breadcrumbs block](https://github.com/woocommerce/woocommerce/blob/3d5e7c3610b543db31f4a3adca39597a49a8e4e7/plugins/woocommerce-blocks/assets/js/blocks/breadcrumbs/index.tsx#L31-L38)?  (More such examples can be found in the codebase)
- Separate file for the icon: I don't think it's necessary here. If we will create a new file for icon then probably we will need to create a new folder for icons in Product Collection block.
- `Icon` from `@wordpress/icons` package: I believe we can use it here as well. Here is how it looks like:
```tsx
import { Icon } from '@wordpress/icons';

<Icon
    icon={ handPickedCollectionIcon }
/>
```

What do you think?

File: plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/hand-picked-products-control.tsx
Comment (imanish003) [reply plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/hand-picked-products-control.tsx]: I am not sure if we should change it based on this. Maybe we can keep just one of these? For example, Maybe we can use "Hand-picked Products" for both cases? But it's worth confirming with Jill 🙂

If we decide to change it, then I think it's okay to pass an extra prop for now. For example if `label` prop is passed then we use it, otherwise we use the default one. Because `HandPickedProductsControl` is just a reusable React component from my POV.

File: plugins/woocommerce-blocks/assets/js/blocks/product-collection/collections/hand-picked.tsx
Comment (imanish003) [inline plugins/woocommerce-blocks/assets/js/blocks/product-collection/collections/hand-picked.tsx]: I know it's still a draft pull request but wanted to mention that we should remove `timeFrame` from the query 🙇‍♂️

File: plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/index.tsx
Comment (imanish003) [inline plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/index.tsx:306]: Nit: I think we can also check for `props.attributes.collection` here instead of inside `CollectionSpecificControls` 🙂

File: plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/hand-picked-products-control.tsx
Comment (WunderBart) [reply plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/hand-picked-products-control.tsx]: > I am not sure if we should change it based on this. Maybe we can keep just one of these? For example, Maybe we can use "Hand-picked Products" for both cases? But it's worth confirming with Jill 

Personally, I'd opt for leaving the `Select Products` if we decide to go with one title (see the screenshots below). No strong opinion here, though. @j111q what do you think? Should we keep one title for both scenarios or go with context-specific ones?

| Hand-Picked collection control | Filters drop-down | Filters form |
| ------------- | ------------- | ------------- |
| Screenshot 2024-09-09 at 16 12 19 | Screenshot 2024-09-09 at 16 12 51 | Screenshot 2024-09-09 at 16 12 59 |

File: plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/index.tsx
Comment (WunderBart) [reply plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/index.tsx:306]: I intentionally moved the current collection check to `CollectionSpecificControls` in case other collections have their own controls in the future so that we're able to selectively render them as we do [here](https://github.com/woocommerce/woocommerce/blob/72112bd5d44918b64344939522139a01fddda556/plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/index.tsx#L164-L190). Does that make sense?

File: plugins/woocommerce-blocks/assets/js/blocks/product-collection/collections/hand-picked.tsx
Comment (WunderBart) [reply plugins/woocommerce-blocks/assets/js/blocks/product-collection/collections/hand-picked.tsx]: > Maybe we can use the same approach as in the [Breadcrumbs block](https://github.com/woocommerce/woocommerce/blob/3d5e7c3610b543db31f4a3adca39597a49a8e4e7/plugins/woocommerce-blocks/assets/js/blocks/breadcrumbs/index.tsx#L31-L38)?

Oh that's perfect - missed this one! Thanks! 🙇

File: plugins/woocommerce-blocks/assets/js/blocks/product-collection/collections/hand-picked.tsx
Comment (WunderBart) [reply plugins/woocommerce-blocks/assets/js/blocks/product-collection/collections/hand-picked.tsx]: Removed in 6e09d7b480985995626ef9c9e5fe1a60df1cea47

File: plugins/woocommerce-blocks/assets/js/blocks/product-collection/collections/hand-picked.tsx
Comment (WunderBart) [reply plugins/woocommerce-blocks/assets/js/blocks/product-collection/collections/hand-picked.tsx]: Updated in de596e3e1197b7a8075598383c33c5d07cb32f2b

Comment (WunderBart) [reply plugins/woocommerce-blocks/assets/js/blocks/product-collection/collections/hand-picked.tsx]: For some reason, the GH syntax highlighting for this file is still broken. 🙄 🤷

Comment (imanish003) [reply plugins/woocommerce-blocks/assets/js/blocks/product-collection/collections/hand-picked.tsx]: Haha IMO As long as the functionality is correct, we're good to go 🤔

File: plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/hand-picked-products-control.tsx
Comment (kmanijak) [reply plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/hand-picked-products-control.tsx]: I'm totally for the unified label. There's no reason to contextualize which just makes the control more complex (code-wise and UX-wise). `Select products` sounds good to me!

Comment (j111q) [reply plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/hand-picked-products-control.tsx]: Hi, it makes sense to me too to use the same label in both contexts 👍 That being said, could we change the copy to `Select specific products` so it's a little clearer? Do you think that makes sense?

Comment (WunderBart) [reply plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/hand-picked-products-control.tsx]: Thanks, @j111q! Looking at the other filters now, I wonder if we should adjust the phrasing for a consistent tone, as `Select specific products` would be the only one that calls to action. What if we go with `Selected Products` or `Products Selection` so that it matches other filters?

Comment (j111q) [reply plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/hand-picked-products-control.tsx]: Oh that's a great point. Maybe we could go with `Hand-picked Products` then for super consistency? 😆  Consistent structure to other filters, and matches the name of the filter directly.

Comment (WunderBart) [reply plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/hand-picked-products-control.tsx]: Following a quick discussion with Jill on Slack, we’ve decided to use `Hand-Picked Products` across both contexts. Additionally, the collection description will be updated to `Select specific products to recommend to customers` to clarify the purpose. 🙌

Comment (WunderBart) [reply plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/hand-picked-products-control.tsx]: Updated in https://github.com/woocommerce/woocommerce/pull/51141/commits/a1a23f32d47b4a0c2057077667918c51323110d7.

**Review by j111q — APPROVED** [review]
copy looks good to me 🙏 thanks bart for the detailed work here!

**Comment by WunderBart** [comment]
@imanish003 @kmanijak @xristos3490, @j111q, what should be the initial state of the collection? Right now, it loads 5 products per the default query/filters. IIUC, because hand-picked is a filter and not a whitelist, it will always render products if the filter value is empty. Should we change that behavior and render `no results` until the user hand-picks something?

Initial state:
Screenshot 2024-09-12 at 12 29 30

--- 

_edit_

As agreed on Slack (p1726136877664979-slack-C05UMTA7W0P), we will display no results (nothing in the front end) until a product is hand-picked.

**Comment by WunderBart** [comment]
This is ready for another look, as all the functionality is now in place! 🙌
I will add tests next week, likely after the conflicts with https://github.com/woocommerce/woocommerce/pull/51076 are resolved.

cc: @imanish003 @kmanijak @xristos3490 @ObliviousHarmony

File: plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php
Comment (kmanijak) [inline plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php]: I'm getting two warnings on the frontend: `Undefined array key "post__in"` from line 879 and 881, I guess we should check if `$acc['post__in']` and `$query['post__in']` exist first.

File: plugins/woocommerce-blocks/assets/js/blocks/product-collection/collections/hand-picked.tsx
Comment (kmanijak) [inline plugins/woocommerce-blocks/assets/js/blocks/product-collection/collections/hand-picked.tsx]: We should hide `CoreFilterNames.FILTERABLE`. Right now we show "Sync with current query" and I'm not sure that's expected:

EDIT: Added `CoreFilterNames.ORDER` to suggestion as per comment elsewhere. That will hide Order control from Inspector Controls.

```suggestion
	hideControls: [ CoreFilterNames.HAND_PICKED, CoreFilterNames.FILTERABLE, CoreFilterNames.ORDER ],
```

File: plugins/woocommerce-blocks/assets/js/blocks/product-collection/collections/hand-picked.tsx
Comment (kmanijak) [inline plugins/woocommerce-blocks/assets/js/blocks/product-collection/collections/hand-picked.tsx]: These are redundant, collection will be found with either of these, just because they are part of collection (variation) and original block name. We can think of other keywords like:

```suggestion
	keywords: [ 'specific', 'manual' ],
```

File: plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/index.tsx
Comment (kmanijak) [inline plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/index.tsx]: Instead of this specific check, please add `CoreFilterNames.ORDER` to `hideControls` in a collection definition.

**Review by kmanijak — DISMISSED** [review]
Amazing work! I'm so glad to see Handpicked Products collection! 🚀 Here's some feedback:

### More than 5 products

I started adding bunch of products and I thought there's some bugs because at some point they didn't appear in collection. And I realized we limit the display to 5 products per page and single page:

image

This setting makes sense for "fixed" collections where we know we want to limit/control the display. For example, we're deliberately not including pagination for other collections, you get 5 products initially, period. In here it's more in user's hands and I'm wondering how should we address this. Some ideas:
1. Set the `perPage` attribute to `0` so we simply display all of the products. We leave it in user's hands to add pagination or limit `perPage` as they wish
2. Increase `perPage` as user adds more products... But that's stupid, it's basically the same as above but more complex and weird, easy to mess with user's expectations 😅 
3. Set the `pages` attribute to `0` and add Pagination block by default.

I'm thinking we should do no 1. for sure. And I'd do 3. as well to increase discoverability (it's not easy to figure out there's Pagination block: that you can add it, where to put it, etc.).

### Default state

I suggest changing the "No results found" to some other text specifically for this collection. Something like "No products chosen yet. Use "Hand-Picked Products" in Inspector Controls to choose products to display". 

For now, I think I'd simply differentiate text [in here](https://github.com/woocommerce/woocommerce/blob/63d93d73551a1cc908e5385818796c9d042eca2f/plugins/woocommerce-blocks/assets/js/blocks/product-template/edit.tsx#L343) depending on the collection type.

image

**Review by kmanijak — COMMENTED** [review]
I made the changes addressing my feedback from previous round of review hence removing "Request changes". But at the same time I don't feel like I should be approving the PR hence would be good to get a review from someone else!

File: plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/hand-picked-products-control.tsx
Comment (imanish003) [inline plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/hand-picked-products-control.tsx]: **Question:** What's the purpose of `collectionSpecific`? I can't find any usage of it in the codebase 🤔

**Review by imanish003 — CHANGES_REQUESTED** [review]
Hey @WunderBart,

I tested the changes and everything seems to be working as expected, except for one issue around pagination.

#### Issue: pagination isn't working

I added products such that pagination is visible on Frontend. But It's not working. I suspect, it's because we are setting `perPage` to `0` and `pages` to `0` [here](https://github.com/woocommerce/woocommerce/blob/4e64665c36964d09244e7adbd28371282d227b4f/plugins/woocommerce-blocks/assets/js/blocks/product-collection/collections/hand-picked.tsx#L65-L66).

**expected behavior:**
When I click on 2nd page, I should see products from 2nd page.

**actual behavior:**
No matter which page I click, I always see products from the 1st page.

Here is a demo video:

https://github.com/user-attachments/assets/7161978c-178c-4362-a756-36e79cad1894

**Comment by imanish003** [comment]
> Issue: pagination isn't working

Fixed pagination issue in 58fd5f3

**Review by imanish003 — APPROVED** [review]

**Comment by imanish003** [comment]
I have approved the PR after:
- Fixing pagination issue
- Updating the PR description
- Adding a changelog entry
- Testing PR again

File: plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/hand-picked-products-control.tsx
Comment (imanish003) [reply plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/hand-picked-products-control.tsx]: CC: @WunderBart

Comment (WunderBart) [reply plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/hand-picked-products-control.tsx]: Ah, nice catch - that's just a leftover from something I tested earlier. Will remove.

File: plugins/woocommerce-blocks/assets/js/blocks/product-template/edit.tsx
Comment (WunderBart) [inline plugins/woocommerce-blocks/assets/js/blocks/product-template/edit.tsx]: Should the product template be tangled up with a specific collection like that? Maybe we could pass this custom 'no results' text via the Hand-Picked collection attributes or sth? @kmanijak @imanish003

Comment (WunderBart) [reply plugins/woocommerce-blocks/assets/js/blocks/product-template/edit.tsx]: After further inspection, I think we should drop this custom 'no results' message entirely. The reason is that it will also show this message when we hand-pick products but then apply filters that clear the results (no intersection).

As an alternative, I suggest improving the wording of the current message so that it's more informative and matches the wider context so that we can use it for all cases:

**Current message** (displayed for all empty collections):

`No results found.`

**Suggested new message**:

_This would also be displayed for all collections. Choose one, but we can mix and match. I think I like the first one the most!_ 😄 

1. `No products to display. Check your filter settings in the block settings panel.`
2. `No products found. Adjust filters in the block settings to populate this collection.`
3. `No products are currently displayed. Try adjusting the filters in the block settings panel.`

> [!IMPORTANT]
> I used the `settings panel` name from the [block editor handbook](https://developer.wordpress.org/block-editor/) nomenclature. I think we should avoid using `inspector controls` as it's not a user-facing name. 

cc: @j111q

Comment (imanish003) [reply plugins/woocommerce-blocks/assets/js/blocks/product-template/edit.tsx]: > The reason is that it will also show this message when we hand-pick products but then apply filters that clear the results (no intersection).

Aha! Good catch.

> As an alternative, I suggest improving the wording of the current message so that it's more informative and matches the wider context so that we can use it for all cases:

I really like your suggestion to improve the general 'no results' message. It provides more context and guidance to the user. Here is my suggestion: "No products found. Try adjusting the filters in the block settings panel if there are products in your store."

- It suggests adjusting filters if there are indeed products in the store that aren't showing up due to the current filter settings.
- It implicitly acknowledges that if there are no products in the store at all, adjusting filters won't help.

What do you think?

@WunderBart As we need to merge this PR on Monday, probably we can go with any of the above. If needed, we can iterate on the message in follow-up PR. what do you think?

Comment (WunderBart) [reply plugins/woocommerce-blocks/assets/js/blocks/product-template/edit.tsx]: Thanks for the input, Manish! Two thoughts on the suggested wording:

- For the first part, I think we should go with `No products **to display**`. It matches both situations where filters are and are not applied, while `No products **found**` implies that a search/filtering has already been performed (misleading in the case of the initial Hand-Picked collection state).
- For the second part, I'd skip the `if there are products in your store` bit because it creates a precedent to be followed in all other product-related blocks. I think it should be implied that products must be available to use product-related blocks.

To sum up, I'd go with:

`No products to display. Try adjusting the filters in the block settings panel.`

WDYT?

Comment (WunderBart) [reply plugins/woocommerce-blocks/assets/js/blocks/product-template/edit.tsx]: > As we need to merge this PR on Monday, probably we can go with any of the above. If needed, we can iterate on the message in follow-up PR. what do you think?

As per the above, I went ahead and updated the copy in 6e2ddf447e53c4eaab3770c6bd603b4100dd3c31.

File: plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/hand-picked-products-control.tsx
Comment (WunderBart) [reply plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/hand-picked-products-control.tsx]: Removed in https://github.com/woocommerce/woocommerce/pull/51141/commits/f8c4a106eeb0b7a0991e88cba77b1f6e9ba15e30.

**Comment by WunderBart** [comment]
The PR tests well for me, @imanish003 @kmanijak. Thanks! 🚀 I'll try adding on some tests here.

File: plugins/woocommerce-blocks/assets/js/blocks/product-template/edit.tsx
Comment (imanish003) [reply plugins/woocommerce-blocks/assets/js/blocks/product-template/edit.tsx]: Makes sense to me. Thanks for updating the copy 🙌🏻

**Comment by kimmenbert** [comment]
Any plans on enhancing this block with custom ordering of handpicked products like mentioned in multiple github issues in the past years. eg: https://github.com/woocommerce/woocommerce/issues/46215#issuecomment-2102760148

**Pull request merged**
